src/docbuilder/metadata.rs

```diff
@@ -43,6 +44,13 @@ pub struct Metadata {
     /// is always built on this target. You can change default target by setting this.
     pub default_target: Option<String>,
 
+    /// If you want a crate to build only for specific targets,
+    /// set `extra-targets` to the list of targets to build, in addition to `default-target`.
+    ///
+    /// If you do not set `extra_targets`, all of the tier 1 supported targets will be built.
+    /// If you set `extra_targets` to an empty array, only the default target will be built.
+    pub extra_targets: Option<Vec<String>>,
```

Comment:
I'd rename this just to `targets`, as it doesn't add *extra* targets, it sets the full list of targets the author wants to build.

Reply:
I don't think this is the case? This specifies targets to build _in addition to_ the default target. The default target will always be built even if `extra-targets` is empty.

Reply:
It's more that it gives me the impression that those targets are *in addition* to the default list of tier 1 targets.

Reply:
Ah, that makes sense ... Do you think that `targets` makes it clear that the default target will still be built no matter what? I'm worried that people will try to set the default target by setting `extra-targets = [default]`

Reply:
Honestly I don't see that as a problem. If no explicit `default-target` is set and `targets` is set we should pick the first item in `targets` as the default.

Reply:
Ok, I can implement that. I think it's better to do the expected thing than to document how to do the unexpected thing :)

Here's what I propose to do:

- If `default-target` is set, use that as the default. If it appears again in `targets`, it is ignored; all other targets are built (removing duplicates).
- If `default-target` is unset and `targets` has at least one target, use the first target as the default and build all others (removing duplicates).
- If `default-target` is unset and `targets` is explicitly set to empty (`[]`), use `DEFAULT_TARGET` (x86-unknown-linux-gnu)
- If `default-target` is unset and `targets` is not set, use `DEFAULT_TARGET` and also build all the tier 1 targets.

Does that sounds reasonable?

Reply:
I don't know ... looking back at this it might be more intuitive, but it's hard to explain. I'd rather go for the simpler version..

Reply:
Writing this a little more concisely for my sanity:

```
targets = [default_target] + targets.unwrap_or(TARGETS);
deduplicate(targets);
if targets.is_empty() { // `default_target` unset, `targets` set
    targets = [DEFAULT_TARGET];
}
landing_page = targets[0];
for target in targets {
    build(target);
}
```

Reply:
This is my code so far, I'm out of energy to work on this. Feel free to take over the PR if you have time :)

```
    pub(super) fn select_extra_targets<'a: 'ret, 'b: 'ret, 'ret>(&'a self, default_target: Option<&'b str>) -> (&'ret str, HashSet<&'ret str>) {
        use super::rustwide_builder::{DEFAULT_TARGET, TARGETS};
        // Let people opt-in to only having specific targets
        let mut all_targets = default_target
            .iter()s
            .chain(self.extra_targets.map_or(TARGETS.iter(), |v| v.iter().map(|s| s.as_str())))
            .copied();
        // default_target unset and extra_targets set to `[]`
        let landing_page = all_targets.next().unwrap_or(DEFAULT_TARGET);
        let mut extras: HashSet<_> = all_targets.collect();
        extras.remove(landing_page);
        (landing_page, extras)
    }
```

Reply:
Implemented this in the most recent commit. I still think it's more complicated than it needs to be, but I'm ok keeping the new behavior if you think it's useful.

Note that this required a significant refactor of the build internals. This hasn't been tested yet, and the PR shouldn't be merged until it has been.